core/src/Streamly/Internal/Data/Unbox/TH.hs

```diff
@@ -0,0 +1,412 @@
+-- function to @n@ and @k@.
+elimTV :: (Name -> r) -> (Name -> Kind -> r) -> TyVarBndr_ flag -> r
+#if MIN_VERSION_template_haskell(2,17,0)
+elimTV ptv _ktv (PlainTV n _)    = ptv n
+elimTV _ptv ktv (KindedTV n _ k) = ktv n k
+#else
+elimTV ptv _ktv (PlainTV n)    = ptv n
+elimTV _ptv ktv (KindedTV n k) = ktv n k
+#endif
+
+-- | Extract the type variable name from a 'TyVarBndr', ignoring the
+-- kind signature if one exists.
+tvName :: TyVarBndr_ flag -> Name
+tvName = elimTV id (\n _ -> n)
+
+-- | Get the 'Name' of a 'TyVarBndr'
+tyVarBndrName :: TyVarBndr_ flag -> Name
+tyVarBndrName = tvName
+
+appsT :: Type -> [Type] -> Type
+appsT x [] = x
+appsT x (y:xs) = appsT (AppT x y) xs
+
+-- | Utility to conveniently handle change to 'InstanceD' API in
+-- template-haskell-2.11.0
+plainInstanceD :: Cxt -> Type -> [Dec] -> Dec
+plainInstanceD =
+#if MIN_VERSION_template_haskell(2,11,0)
+    InstanceD Nothing
+#else
+    InstanceD
+#endif
+
+-- | Simplified info about a 'DataD'. Omits deriving, strictness,
+-- kind info, and whether it's @data@ or @newtype@.
+data DataType = DataType
+    { dtName :: Name
+    , dtTvs :: [Name]
+    , dtCxt :: Cxt
+    , dtCons :: [DataCon]
+    } deriving (Eq, Show, Ord) --, Data, Typeable, Generic)
+
+-- | Simplified info about a 'Con'. Omits deriving, strictness, and kind
+-- info. This is much nicer than consuming 'Con' directly, because it
+-- unifies all the constructors into one.
+data DataCon = DataCon
+    { dcName :: Name
+    , dcTvs :: [Name]
+    , dcCxt :: Cxt
+    , dcFields :: [(Maybe Name, Type)]
+    } deriving (Eq, Show, Ord) --, Data, Typeable, Generic)
+
+
+-- | Convert a 'Con' to a list of 'DataCon'. The result is a list
+-- because 'GadtC' and 'RecGadtC' can define multiple constructors.
+conToDataCons :: Con -> [DataCon]
+conToDataCons = \case
+    NormalC name slots ->
+        [DataCon name [] [] (map (\(_, ty) -> (Nothing, ty)) slots)]
+    RecC name fields ->
+        [DataCon name [] [] (map (\(n, _, ty) -> (Just n, ty)) fields)]
+    InfixC (_, ty1) name (_, ty2) ->
+        [DataCon name [] [] [(Nothing, ty1), (Nothing, ty2)]]
+    ForallC tvs preds con ->
+        map (\(DataCon name tvs0 preds0 fields) ->
+            DataCon name (tvs0 ++ map tyVarBndrName tvs) (preds0 ++ preds) fields) (conToDataCons con)
+#if MIN_VERSION_template_haskell(2,11,0)
+    GadtC ns slots _ ->
+        map (\dn -> DataCon dn [] [] (map (\(_, ty) -> (Nothing, ty)) slots)) ns
+    RecGadtC ns fields _ ->
+        map (\dn -> DataCon dn [] [] (map (\(fn, _, ty) -> (Just fn, ty)) fields)) ns
+#endif
+
+-- | Reify the given data or newtype declaration, and yields its
+-- 'DataType' representation.
+reifyDataType :: Name -> Q DataType
+reifyDataType name = do
+    info <- reify name
+    case infoToDataType info of
+        Nothing -> fail $ "Expected to reify a datatype. Instead got:\n" ++ pprint info
+        Just x -> return x
+
+
+infoToDataType :: Info -> Maybe DataType
+infoToDataType info = case info of
```

Comment:
Always use increasing indentation.

Reply:
I'm not changing the code I've copied. I'm keeping it like how it is.